google-http-client-android/src/main/java/com/google/api/client/extensions/android/util/store/FileDataStoreFactory.java

```diff
@@ -0,0 +1,161 @@
+  static class FileDataStore<V extends Serializable> extends AbstractMemoryDataStore<V> {
+
+    /** File to store data. */
+    private final File dataFile;
+
+    FileDataStore(FileDataStoreFactory dataStore, File dataDirectory, String id)
+        throws IOException {
+      super(dataStore, id);
+      this.dataFile = new File(dataDirectory, id);
+      // error if it is a symbolic link
+      if (IOUtils.isSymbolicLink(dataFile)) {
+        throw new IOException("unable to use a symbolic link: " + dataFile);
+      }
+      // create new file (if necessary)
+      if (dataFile.createNewFile()) {
+        keyValueMap = Maps.newHashMap();
+        // save the credentials to create a new file
+        save();
+      } else {
+        // load credentials from existing file
+        keyValueMap = IOUtils.deserialize(new FileInputStream(dataFile));
+      }
+    }
+
+    @Override
+    public void save() throws IOException {
+      IOUtils.serialize(keyValueMap, new FileOutputStream(dataFile));
+    }
+
+    @Override
+    public FileDataStoreFactory getDataStoreFactory() {
+      return (FileDataStoreFactory) super.getDataStoreFactory();
+    }
+  }
+
+  /**
+   * Attempts to set the given file's permissions such that it can only be read, written, and
+   * executed by the file's owner.
+   *
+   * @param file the file's permissions to modify
+   * @throws IOException
+   */
+  static void setPermissionsToOwnerOnly(File file) throws IOException {
+    // Disable access by other users if O/S allows it and set file permissions to readable and
+    // writable by user. Use reflection since JDK 1.5 will not have these methods
+    try {
+      Method setReadable = File.class.getMethod("setReadable", boolean.class, boolean.class);
+      Method setWritable = File.class.getMethod("setWritable", boolean.class, boolean.class);
+      Method setExecutable = File.class.getMethod("setExecutable", boolean.class, boolean.class);
+      if (!(Boolean) setReadable.invoke(file, false, false)
+          || !(Boolean) setWritable.invoke(file, false, false)
+          || !(Boolean) setExecutable.invoke(file, false, false)) {
+        LOGGER.warning("unable to change permissions for everybody: " + file);
+      }
+      if (!(Boolean) setReadable.invoke(file, true, true)
+          || !(Boolean) setWritable.invoke(file, true, true)
+          || !(Boolean) setExecutable.invoke(file, true, true)) {
+        LOGGER.warning("unable to change permissions for owner: " + file);
+      }
+    } catch (InvocationTargetException exception) {
+      Throwable cause = exception.getCause();
+      Throwables.propagateIfPossible(cause, IOException.class);
+      // shouldn't reach this point, but just in case...
+      throw new RuntimeException(cause);
+    } catch (NoSuchMethodException exception) {
+      LOGGER.warning(
+          "Unable to set permissions for "
+              + file
+              + ", likely because you are running a version of Java prior to 1.6");
```

Comment:
Is this the right error message here?

Reply:
Probably not, this was the version from google-http-client (not restricted to android). I'll see if I can come up with a better message.

Reply:
Removed the reflection as the methods we were previously protecting were added in [API level 9](https://developer.android.com/reference/java/io/File#setReadable(boolean,%2520boolean)).